lua/frecency/db.lua

```diff
@@ -0,0 +1,186 @@
+---@field auto_validate boolean: When this to false, stale entries will never be automatically removed, default true
+db.config = {
+  db_root = nil,
+  ignore_patterns = {},
+  safe_mode = true,
+  auto_validate = true,
+}
+
+---Set database configuration
+---@param config FrecencyDBConfig
+db.set_config = function(config)
+  db.config = vim.tbl_extend("keep", config, db.config)
+end
+
+---Initialize frecency Database
+---@param config FrecencyDBConfig
+---@overload func()
+---@return FrecencyDB
+db.init = function(config)
+  config = config or {}
+  db.db.uri = config.db_root and config.db_root or db.db.uri
+  db.config = config and db.set_config(config) or db.config
+  db.db:init()
+  db.is_initialized = db.db.is_initialized ---TODO: remove when sql.nvim@#93 is fixed
+
+  ---Use oldfiles on first run.
+  if db.files:count() == 0 then
+    -- TODO: this needs to be scheduled for after shada load??
+    local oldfiles = vim.api.nvim_get_vvar "oldfiles"
+    for _, path in pairs(oldfiles) do
+      db.files.insert { path = path }
+    end
+    print(("Telescope-Frecency: Imported %d entries from oldfiles."):format(#oldfiles))
+  end
+end
+
+---Get timestamps with a computed filed called age.
+---If file_id is nil, then get all timestamps.
+---@param opts table
+---- { file_id } number: id file_id corresponding to `files.id`. return all if { file_id } is nil
+---- { with_age } boolean: whether to include age, default false.
+---@return table { id, file_id, age }
+---@overload func()
+db.get_timestamps = function(opts)
+  opts = opts or {}
+  local where = opts.file_id and { file_id = opts.file_id } or nil
+  local compute_age = opts.with_age and s.cast((s.julianday() - s.julianday "timestamp") * 24 * 60, "integer") or nil
+  return db.timestamps.get { where = where, keys = { age = compute_age, "id", "file_id" } }
+end
+
+---Get file entries
+---@param opts table:
+---- { ws_path } string: get files with matching workspace path.
+---- { show_unindexed } boolean: whether to include unindexed files, false if no ws_path is given.
+---- { with_score } boolean: whether to include score in the result and sort the files by score.
+---@overload func()
+---@return table[]: files entries
+db.get_files = function(opts)
+  opts = opts or {}
+  local contains = opts.ws_path and { path = { opts.ws_path .. "*" } } or nil
+  local files = db.files.get { contains = contains }
+
+  if opts.ws_path and opts.show_unindexed then
+    util.include_unindexed(files, opts.ws_path)
+  end
+
+  if opts.with_score then
+    ---NOTE: this might get slower with big db, it might be better to query with db.get_timestamp.
+    ---TODO: test the above assumption
+    local timestamps = db.get_timestamps { with_age = true }
+    for _, file in ipairs(files) do
+      file.timestamps = util.tbl_match("file_id", file.id, timestamps)
+      file.score = algo.calculate_file_score(file)
+    end
+    table.sort(files, function(a, b)
+      return a.score > b.score
+    end)
+  end
```

Comment:
Does it make sense to first do `with_score` and do the sorting and afterwards potentially append the `unindexed`. They all have a score of 0 so it makes no sense to sort with them already in mind. Or am i missing something?

Reply:
👍 good catch! makes prefect sense